src/http-libwebsockets.c

```diff
@@ -673,77 +732,106 @@ static int add_header_by_name(struct lws *wsi, const char* name, const char* val
+            // alloc_pool metrics (+ 1 for alloc_pool_total_bytes):
+            + (DEQ_SIZE(allocator_metrics) * PER_METRIC_BUF_SIZE * PER_ALLOC_METRIC_COUNT)
+            + PER_METRIC_BUF_SIZE
+            // 1 terminating null
+            + 1;
+        stats->state = new_stats_request_state(buf_size);
+        stats->state->wsi = wsi;
+        stats->state->server = hs;
         //request stats from core thread
-        qdr_request_global_stats(hs->core, &stats->context->stats, handle_stats_results, (void*) stats->context);
+        qdr_request_global_stats(hs->core, &stats->state->stats, handle_stats_results, (void*) stats->state);
         return 0;
     }
 
     case LWS_CALLBACK_HTTP_WRITEABLE: {
-        //encode stats into buffer
-        if (!stats->headers_sent) {
-            if (lws_add_http_header_status(wsi, HTTP_STATUS_OK, &position, end)
-                || add_header_by_name(wsi, "content-type:", "text/plain", &position, end)
-                || add_header_by_name(wsi, "connection:", "close", &position, end))
-                return 1;
-            if (lws_finalize_http_header(wsi, &position, end))
-                return 1;
-            stats->headers_sent = true;
+        // LWS HTTP server ready to send to HTTP response data
+        assert(stats->state);  // expect LWS_CALLBACK_HTTP event occurs first!
+
+        if (stats->response_complete) {  // ignore spurious WRITABLE events once response complete
+            return 0;
         }
 
-        while (stats->current < metrics_length) {
-            if (write_metric(&position, end, &metrics[stats->current], &stats->context->stats)) {
-                stats->current++;
-                qd_log(LOG_HTTP, QD_LOG_DEBUG, "wrote metric %lu of %lu", stats->current, metrics_length);
-            } else {
-                qd_log(LOG_HTTP, QD_LOG_WARNING, "insufficient space in buffer");
-                break;
-            }
+        if (!stats->state->callback_completed) {
+            // the asynchronous request for global metrics has not yet completed. When it does another
+            // LWS_CALLBACK_HTTP_WRITABLE event will be generated and then we can send the response.
+            return 0;
         }
 
-        int alloc_cur = 0;
-        while (alloc_cur < allocator_metrics_length) {
-            if (write_allocator_metric(&position, end, &allocator_metrics[alloc_cur])) {
-                qd_log(LOG_HTTP, QD_LOG_DEBUG, "wrote allocator metric %i of %lu", alloc_cur,
-                       allocator_metrics_length);
-                alloc_cur++;
-            } else {
-                qd_log(LOG_HTTP, QD_LOG_WARNING, "insufficient space in buffer");
-                break;
-            }
+        uint8_t *start = &stats->state->lws_prefix[LWS_PRE];
+        uint8_t *end = start + stats->state->buffer_size;  // first byte past buffer
+
+        // encode stats into buffer
+
+        if (lws_add_http_header_status(wsi, HTTP_STATUS_OK, &start, end)
```

Comment:
Can we please add a comment here.
Looking at the code, I am assuming that it is trying to first set the response header status to 200, sets the content type and text/plain and adds a connection: header and  finally finishes off the http header and if *any* of these operations fail, the whole request fails.

Reply:
> Can we please add a comment here.

Don't worry about this - the log message `Metrics request failed: cannot send headers` is good enough